src/rng.rs

```diff
@@ -14,7 +14,43 @@ use crate::distributions::uniform::{SampleUniform, UniformSampler};
 use crate::distributions::{self, Distribution, Standard};
 use core::num::Wrapping;
 use core::{mem, slice};
-use core::ops::RangeBounds;
+use core::ops::{Range, RangeInclusive};
+
+/// Range that supports generating a single sample efficiently.
+///
+/// Any type implementing this trait can be used to specify the sampled range
+/// for `Rng::gen_range`.
+pub trait SampleRange<T> {
+    /// Generate a sample from the given range.
+    fn sample_single<R: RngCore + ?Sized>(self, rng: &mut R) -> T;
+
+    /// Check whether the range is empty.
+    fn is_empty(&self) -> bool;
+}
+
+impl<T: SampleUniform + PartialOrd> SampleRange<T> for Range<T> {
+    #[inline]
+    fn sample_single<R: RngCore + ?Sized>(self, rng: &mut R) -> T {
+        T::Sampler::sample_single(self.start, self.end, rng)
+    }
+
+    #[inline]
+    fn is_empty(&self) -> bool {
+        !(self.start < self.end)
```

Comment:
Since integers can't have NaN values, there's no need for the negation here. :shrug:

Reply:
We also support float ranges!

Reply:
Ach, me forgetting that ranges aren't only used with `usize`!